Fix error : Uncaught TypeError: cb is not a function

alain75007 opened the pull request:
Mongoose: 4.2.4
Node: 5.0.0

pdesterlich commented:
I am some kind of newbie in node development so it may be my fault, but can you please explain me what's the point in your changes? Because all i can see is that you add an argument to a function but you never use it.

I updated my environment to match node & mongoose versions, but i get no error when running the included tests with my unmodified code; i also ran tests in a node app i'm working on (with lots of models relying on this plugin) with updated mongoose but i still get no error.
Can you provide me some details on how to reproduce, so i can update my tests?

alain75007 commented:
Hi, thanks to give some attention to my bug. In some case the method *schema.methods.save* is called with 2 arguments, causing the bug.  In this case, the first argument always contains this : * { __noPromise: true }*.  In don't know why...  
The patch in the pull request is not sufficient to fix the bug, I replaced it with this : 
```
  schema.methods.save = function(foo, cb){
    if (!cb) cb = foo;
```

My schema look like this:
````
var schema = new Schema({
  atype: {type: String, required: true, default: "foo"},
  slug: {
    type: String,
    indexed: true,
    sparse: false,
    unique: true,
    required: true
  }
},   discriminatorKey: 'access_type');

var My =  MyModel.discriminator('foo', schema);

```
And I declare the plugin like this:
```
MySchema.plugin(slugin, {
  source: 'textForSlug'
});
```
*TextForSlug* is a virtual path.

Commit:
Fix error : Uncaught TypeError: cb is not a function

Mongoose: 4.2.4
Node: 5.0.0

In some case the method schema.methods.save is called with 2 arguments,
causing the bug. In this case, the first argument always contains this :
 { __noPromise: true }.

pdesterlich commented:
uhm ... are you using some other plugins on the same model that implement the same schema.methods.save as mine?

Commit:
Add two tests (which produce errors)

alain75007 commented:
Ok, the first cb error happen when I use Create instead of Save.
The second error (duplicate key) is more mysterious.
